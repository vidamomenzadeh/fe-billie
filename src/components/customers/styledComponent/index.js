import styled from 'styled-components'

export const BurronWrapper = styled.div`
  margin-top: 20px;
  button {
    float: right;
  }
`
