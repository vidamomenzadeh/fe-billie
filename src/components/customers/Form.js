import {useRef, useState} from 'react'
import {connect} from 'react-redux'

import {makeStyles} from '@material-ui/core/styles'
import Modal from '@material-ui/core/Modal'
import Backdrop from '@material-ui/core/Backdrop'
import Fade from '@material-ui/core/Fade'
import TextField from '@material-ui/core/TextField'
import Button from '@material-ui/core/Button'
import Dialog from '@material-ui/core/Dialog'
import DialogActions from '@material-ui/core/DialogActions'
import DialogContent from '@material-ui/core/DialogContent'
import DialogContentText from '@material-ui/core/DialogContentText'
import DialogTitle from '@material-ui/core/DialogTitle'

import {updateCustomer} from '../../actions/customer'

import {BurronWrapper} from '../customers/styledComponent'

const useStyles = makeStyles((theme) => ({
  modal: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center'
  },
  paper: {
    backgroundColor: theme.palette.background.paper,
    padding: theme.spacing(2, 4, 3),
    minWidth: '600px'
  }
}))

export const UpdateBudget = (props) => {
  const classes = useStyles()
  const inputEl = useRef(null)
  const [valid, setValid] = useState(true)
  const [showDialog, setShowDialog] = useState(false)
  const {open, handleClose, selectedCustomer = {}, updateCustomer} = props

  const isValid = () => {
    if (parseFloat(inputEl.current.value) < selectedCustomer.budget_spent) {
      setValid(false)

      return
    }

    setValid(true)
  }

  const submit = () => {
    updateCustomer({
      customer: {
        ...selectedCustomer,
        budget: parseFloat(inputEl.current.value)
      }
    })

    setShowDialog(true)
  }

  const hideDialog = () => {
    setShowDialog(false)
    handleClose()
  }

  return (
    <div>
      <Modal
        aria-labelledby='transition-modal-title'
        aria-describedby='transition-modal-description'
        data-testid='update-modal'
        className={classes.modal}
        open={open}
        onClose={handleClose}
        closeAfterTransition
        BackdropComponent={Backdrop}
        BackdropProps={{
          timeout: 500
        }}
      >
        <Fade in={open}>
          <div className={classes.paper}>
            <h2 id='transition-modal-title'>{selectedCustomer.name}</h2>
            <TextField
              required
              inputRef={inputEl}
              type='number'
              id='standard-required'
              label='Budget'
              defaultValue={selectedCustomer.budget}
              onChange={isValid}
              data-testid='input-budget-value'
            />
            <BurronWrapper>
              <Button
                variant='contained'
                color='primary'
                disabled={!valid}
                onClick={submit}
                data-testid='btn-submit'
              >
                Submit
              </Button>
            </BurronWrapper>
          </div>
        </Fade>
      </Modal>
      <Dialog
        open={showDialog}
        onClose={handleClose}
        fullWidth={true}
        aria-labelledby='responsive-dialog-title'
      >
        <DialogTitle id='responsive-dialog-title'>
          {'Budget is updated'}
        </DialogTitle>
        <DialogContent>
          <DialogContentText>Go back to List</DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button
            autoFocus
            onClick={() => setShowDialog(false)}
            color='primary'
          >
            Cancel
          </Button>
          <Button onClick={hideDialog} color='primary' autoFocus>
            Ok
          </Button>
        </DialogActions>
      </Dialog>
    </div>
  )
}

export default connect(null, {
  updateCustomer
})(UpdateBudget)
