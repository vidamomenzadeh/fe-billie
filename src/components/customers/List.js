import {useEffect, useState, Fragment} from 'react'
import {connect} from 'react-redux'

import {fetchCustomer} from '../../actions/customer'
import {getCustomerById} from '../../reducers/customer'

import Form from './Form'
import {toGermanFormat} from '../../utils'

import {makeStyles} from '@material-ui/core/styles'
import Table from '@material-ui/core/Table'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableContainer from '@material-ui/core/TableContainer'
import TableHead from '@material-ui/core/TableHead'
import TableRow from '@material-ui/core/TableRow'
import Paper from '@material-ui/core/Paper'

const useStyles = makeStyles({
  table: {
    minWidth: 650
  },
  tableHeader: {
    background: '#ff4338'
  }
})

const ListOfCustomers = ({fetchCustomer, customers}) => {
  const classes = useStyles()
  const [showModal, setShowModal] = useState(false)
  const [selectedCustomer, setSelectedCustomer] = useState({})

  useEffect(() => {
    fetchCustomer()
  }, [])

  const handleClose = () => setShowModal(false)

  const renderCustomersRows = (customers) =>
    Object.keys(customers).map((row) => {
      const customer = customers[row]

      return (
        <TableRow
          key={`customer-${customer.id}`}
          onClick={() => {
            setShowModal(true)
            setSelectedCustomer(customer)
          }}
        >
          <TableCell component='th' scope='row'>
            {customer.name}
          </TableCell>

          <TableCell align='center'>
            {customer.date_of_first_purchase}
          </TableCell>
          <TableCell align='center'>
            {toGermanFormat(customer.budget)}
          </TableCell>
          <TableCell align='center'>
            {toGermanFormat(customer.budget_spent)}
          </TableCell>
          <TableCell align='center'>
            {toGermanFormat(customer.budget_left)}
          </TableCell>
        </TableRow>
      )
    })

  return (
    <Fragment>
      <TableContainer component={Paper}>
        <Table className={classes.table} aria-label='simple table'>
          <TableHead>
            <TableRow className={classes.tableHeader}>
              <TableCell>Company Name</TableCell>
              <TableCell align='center'>Date of first purchase</TableCell>
              <TableCell align='center'>Total budget</TableCell>
              <TableCell align='center'>Budget spent</TableCell>
              <TableCell align='center'>Budget left</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>{renderCustomersRows(customers)}</TableBody>
        </Table>
      </TableContainer>
      <Form
        open={showModal}
        handleClose={handleClose}
        selectedCustomer={selectedCustomer}
      />
    </Fragment>
  )
}

function mapStateToProps(state) {
  return {
    customers: getCustomerById(state)
  }
}

export default connect(mapStateToProps, {
  fetchCustomer
})(ListOfCustomers)
