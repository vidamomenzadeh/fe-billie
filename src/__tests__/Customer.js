/* eslint-disable no-undef */
import {UpdateBudget} from '../components/customers/Form'
import {render, fireEvent} from '@testing-library/react'

const TEST_CUSTOMER = {
  id: 2,
  name: 'Solar Firma',
  budget: 1123.22,
  budget_spent: 451.3754,
  date_of_first_purchase: '2120 - 01 - 14'
}

test('Customer modal render correctly when props open is true', async () => {
  const {container, getByTestId} = render(<UpdateBudget open={true} />)

  expect(getByTestId('update-modal')).toBeInTheDocument()
  expect(getByTestId('input-budget-value')).toBeInTheDocument()
  expect(getByTestId('btn-submit')).toBeInTheDocument()
  expect(container).toMatchSnapshot()
})

test('Customer modal render correctly when props open is false', async () => {
  const {container, queryAllByTestId} = render(<UpdateBudget open={false} />)

  expect(queryAllByTestId('update-modal')).toHaveLength(0)
  expect(queryAllByTestId('input-budget-value')).toHaveLength(0)
  expect(queryAllByTestId('btn-submit')).toHaveLength(0)

  expect(container).toMatchSnapshot()
})

test('submit butn should be disabled when the input value is less than spend value', async () => {
  const {getByTestId} = render(
    <UpdateBudget open={true} selectedCustomer={TEST_CUSTOMER} />
  )
  const inputBudget = getByTestId('input-budget-value').querySelector('input')
  const btnSubmit = getByTestId('btn-submit')

  await fireEvent.change(inputBudget, {target: {value: '90'}})

  expect(btnSubmit).toBeDisabled()
})

test('submit butn should be disabled when the input value is bigger than spend value', async () => {
  const {getByTestId} = render(
    <UpdateBudget open={true} selectedCustomer={TEST_CUSTOMER} />
  )
  const inputBudget = getByTestId('input-budget-value').querySelector('input')
  const btnSubmit = getByTestId('btn-submit')

  await fireEvent.change(inputBudget, {target: {value: '9000'}})

  expect(btnSubmit).not.toBeDisabled()
})
