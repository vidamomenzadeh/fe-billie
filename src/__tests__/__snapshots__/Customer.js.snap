// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customer modal render correctly when props open is false 1`] = `
<div>
  <div />
</div>
`;

exports[`Customer modal render correctly when props open is true 1`] = `
<div
  aria-hidden="true"
>
  <div />
</div>
`;
